woocommerce/woocommerce-gateway-stripe
# Remove intent check from charge succeeded webhook

# Changes proposed in this Pull Request:

Fixes #1850 

The original issue pointed out that SEPA payments made using UPE were not going through the proper process of going to `on-hold` then finally into `completed` / `processing` once the charge has succeeded. The first half of this issue (moving the status to on-hold) was completed as part of #1780. This PR addresses the second half. 

[SEPA payments are not completed immediately](https://stripe.com/payments/payment-methods-guide#sepa-direct-debit) and thus require a webhook to confirm a successful payment. There were two possible hooks to listen for, `charge.succeeded` or `payment_intent.succeeded`. The charge event was getting [blocked from processing since there was a check for a payment intent](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/8eba0d6176c40107c12c2ebd0cfe362549bbb18d/includes/class-wc-stripe-webhook-handler.php#L427-L429) on the charge and if there was, it would not process it. The apparent reason for this check was to avoid any duplicate processing on the order. The payment intent was getting [blocked from processing since the order status was not pending or failed](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/8eba0d6176c40107c12c2ebd0cfe362549bbb18d/includes/class-wc-stripe-webhook-handler.php#L738-L745). With both hooks not getting processed the orders would remain `on-hold`.

With this PR I decided to address the issue by unblocking the charge event by removing the check for a payment intent. There were a few reasons I decided on that option:

- The non-UPE SEPA payments are confirmed using the `charge.succeeded` hook, so with this change the same event is used to confirm SEPA payments regardless of UPE or not.
- The payment intent check appeared to be in place to avoid duplication, but there is an additional check on the order status in both handlers that do not overlap by default. The charge hook will only [process orders on-hold](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/8eba0d6176c40107c12c2ebd0cfe362549bbb18d/includes/class-wc-stripe-webhook-handler.php#L443-L445) and the payment intent hook will only [process orders in pending or failed](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/8eba0d6176c40107c12c2ebd0cfe362549bbb18d/includes/class-wc-stripe-webhook-handler.php#L738-L745) (subject to a custom filter result). In addition, I ran a test where I forced both hooks to process and the only repercussion I saw was a doubling of the order notes, otherwise the WC Orders were updated properly and the Stripe payment was unaffected.
- The `charge.failed` hook has no similar check against a payment intent existing or not.
- Given the very narrow focus of the charge hook only working on orders with the on-hold status, it seemed prudent to keep this change as narrow as possible given its impact on UPE and non-UPE payments.
- Before suffering from any more analysis paralysis I had to pick one approach over the other 😄 

# Testing instructions

Since this PR is focused around webhooks, make sure to be listening for any webhooks throughout all the testing of this PR. At times it has been found that the webhooks stop getting through, in which case restart the listen process and it tends to fix this issue.

- Install [Stripe CLI](https://stripe.com/docs/stripe-cli).
- Run `stripe login` and select the Stripe account you're connected to.
- Run `stripe listen --forward-to 'http://localhost:8082/?wc-api=wc_stripe'` and copy the webhook signing secret (`whsec_******`) to the plugin settings.

With the webhook listener running, the following tests can be run:

- Start with this branch pulled down and UPE enabled with the SEPA payment method enabled within UPE.
- As a customer go to checkout with a product and pay with the SEPA payment method using the following IBAN: `AT321904300235473204`. This IBAN will send a `charge.succeeded` hook after 3 minutes.
- As the merchant go to check on the new order and confirm it has the status On-Hold and that it has order notes marking there is a charge awaiting payment.
- After 3 minutes a `charge.succeeded` webhook should be received. 
- As the merchant check the order again and the order status should be Completed or Processing and two order notes should be added to mark the status update and the charge complete.
- In addition the order details should have a link to the Stripe payment and this can be reviewed as well that the payment was handled successfully.

At this point, the issues from #1850 should all be addressed, but since this change could impact other flows this PR should get some additional regression testing. Things to call out specifically:
- Using other [SEPA IBANs](https://stripe.com/docs/testing#sepa-direct-debit), in particular a failing test with `AT051904300235473205`
- Using SEPA with non-UPE
- Using the Pay for Order page to checkout
- Using a 3DS card that requires authentication

-------------------
